### PR TITLE
WIP: Create Associations between Speakers and Meetings

### PR DESCRIPTION
**Pull request by croz1007**

### CODE REVIEW

File: lib/cbus_elixir/app/meeting.ex [hunk 1]
```diff
@@ -6,6 +6,7 @@ defmodule CbusElixir.App.Meeting do
 
   schema "meetings" do
     field :date, :utc_datetime
+    has_many :speaker, CbusElixir.App.Speaker
```
Comment (mdarby) [inline lib/cbus_elixir/app/meeting.ex]: I *think* it's `speakers`

Comment (croz1007) [reply lib/cbus_elixir/app/meeting.ex]: Updated

**Comment by croz1007** [comment]
@mdarby Can you review this new commit? Herman and I got the future meeting dates displaying on the New Speaker Form. It is inserting the correct info to the Speaker table. Thanks

**Follow ups** 

- Update the Index and Show pages to show the Date of the Meeting and not the Meeting ID.

File: lib/cbus_elixir_web/controllers/speaker_controller.ex [hunk 2]
```diff
@@ -10,8 +12,14 @@ defmodule CbusElixirWeb.SpeakerController do
   end
 
   def new(conn, _params) do
+    today = Timex.today()
+    meets = CbusElixir.Repo.all(from m in "meetings", 
+      where: m.date > type(^today, Ecto.Date),
+      limit: 5,
+      select: {m.id, m.date})
+    meetings = for {k, v} <- meets, do: {Timex.to_date(v) |> Date.to_string, k}
```
Comment (al2o3cr) [inline lib/cbus_elixir_web/controllers/speaker_controller.ex]: I need to read up on the idiomatic way to do this in Phoenix, but at first glance it feels like some of this belongs in a helperish kind of function ("give me the next N upcoming meetings" or similar).

Comment (al2o3cr) [reply lib/cbus_elixir_web/controllers/speaker_controller.ex]: Also, unless there's a default `ORDER BY` getting smuggled into the query from elsewhere the one above has a bug...

**Review by al2o3cr — DISMISSED** [review]
Check into whether the query to load the next 5 meetings gets an `ORDER BY` clause.

File: lib/cbus_elixir_web/controllers/page_controller.ex [hunk 3]
```diff
@@ -1,11 +1,26 @@
...
 
   alias CbusElixir.Repo
   alias CbusElixir.App.Speaker
+  alias CbusElixir.App.Meeting
 
   def index(conn, _params) do
-    speakers = Repo.all(Speaker)
-    render conn, "index.html", speakers: speakers
+    today = Timex.today()
+    meeting = Repo.all(from m in Meeting,
```
Comment (al2o3cr) [inline lib/cbus_elixir_web/controllers/page_controller.ex]: This seems similar to calling `ActiveRecord::Base.connection.select_all` back in Rails - a low-level interface that skips most of the mapping layer.

I recommend [this blog post](http://blog.plataformatec.com.br/2015/08/working-with-ecto-associations-and-embeds/) for more about Ecto associations. We should use them. :)

File: lib/cbus_elixir_web/controllers/page_controller.ex [hunk 4]
```diff
@@ -1,11 +1,26 @@
 defmodule CbusElixirWeb.PageController do
+  import Ecto.Query, only: [from: 2]
```
Comment (al2o3cr) [inline lib/cbus_elixir_web/controllers/page_controller.ex]: This seems like a smell that model behavior is in the wrong place.

**Comment by croz1007** [comment]
@al2o3cr @mdarby 
Can you please pull and review?

File: lib/cbus_elixir_web/controllers/page_controller.ex (hunk 4 above)
Comment (mdarby) [reply lib/cbus_elixir_web/controllers/page_controller.ex]: Yeah, L9:16 should go into PageView as a function that is called from either the controller or the template.

File: lib/cbus_elixir_web/controllers/speaker_controller.ex (hunk 2 above)
Comment (mdarby) [reply lib/cbus_elixir_web/controllers/speaker_controller.ex]: Same thing with PageView here; I'd love to see the `get_meetings` function moved out of the controller and into the view.

Comment (croz1007) [reply lib/cbus_elixir_web/controllers/speaker_controller.ex]: So you are saying to just completely lift it up and plunk it into the `speaker_view.ex` file? Will the controller still pick up the call?

File: lib/cbus_elixir_web/controllers/page_controller.ex (hunk 4 above)
Comment (croz1007) [reply lib/cbus_elixir_web/controllers/page_controller.ex]: Sure...have to figure out how.

Comment (croz1007) [reply lib/cbus_elixir_web/controllers/page_controller.ex]: Okay, I moved the function to the PageView and commented out the function on the controller. Now it cannot see the function. Aren't there already wired up to talk to each other?

**Comment by croz1007** [comment]
@mdarby I moved the `get_meetings` function from the SpeakerController to the SpeakerView. This is working. Trying to figure out the Page level one. Please review my update. Thanks

File: lib/cbus_elixir_web/controllers/page_controller.ex (hunk 4 above)
Comment (al2o3cr) [reply lib/cbus_elixir_web/controllers/page_controller.ex]: Helpers is not where it goes either, IMO - this should be done in `CbusElixir.App.Meeting`. Something like:
```elixir
  def next_meeting_for(today) do
    Repo.one!(from m in Meeting,
      where: m.date > type(^today, Ecto.Date),
      order_by: [asc: m.date],
      preload: [:speakers],
      limit: 1)
  end
```

Parts of this query should likely be extracted to separate functions on `Meeting` composed with `|>`, but that's a separate issue.

Comment (al2o3cr) [reply lib/cbus_elixir_web/controllers/page_controller.ex]: I need to do some more reading on this - "Programming Phoenix" suggests keeping query-building in the models but interactions with `Repo` in the controller (Chapter 7)

File: lib/cbus_elixir_web/controllers/speaker_controller.ex [hunk 5]
```diff
@@ -57,4 +58,13 @@ defmodule CbusElixirWeb.SpeakerController do
     |> put_flash(:info, "Speaker deleted successfully.")
     |> redirect(to: speaker_path(conn, :index))
   end
+
+  # def get_meetings do
```
Comment (mdarby) [inline lib/cbus_elixir_web/controllers/speaker_controller.ex]: any reason for the commented out code?

File: lib/cbus_elixir_web/templates/page/index.html.eex [hunk 6]
```diff
@@ -31,7 +31,7 @@
         <div class="next-meeting">
             <h2 class="title myriad">Next Meeting</h2>
             <p>
-                BLAH BLAH at 6pm<br />
+                <%= formated_date(get_next_meeting()) %> at 6pm<br />
```
Comment (mdarby) [inline lib/cbus_elixir_web/templates/page/index.html.eex]: `formatted` :D

File: lib/cbus_elixir_web/views/page_view.ex [hunk 7]
```diff
@@ -15,7 +19,32 @@ defmodule CbusElixirWeb.PageView do
     "https://www.gravatar.com/avatar/#{hash}"
   end
 
+  def get_next_meeting() do
```
Comment (mdarby) [inline lib/cbus_elixir_web/views/page_view.ex]: This is nitpicky; but `get_` is a smell too :)

**Review by mdarby — CHANGES_REQUESTED** [review]
So very close! :)

File: lib/cbus_elixir_web/controllers/speaker_controller.ex (hunk 5 above)
Comment (croz1007) [reply lib/cbus_elixir_web/controllers/speaker_controller.ex]: That was the original place of the method you wanted me to move. I will remove it.

File: lib/cbus_elixir_web/templates/page/index.html.eex (hunk 6 above)
Comment (croz1007) [reply lib/cbus_elixir_web/templates/page/index.html.eex]: Fixed.

File: lib/cbus_elixir_web/views/page_view.ex (hunk 7 above)
Comment (croz1007) [reply lib/cbus_elixir_web/views/page_view.ex]: Fixed

**Comment by croz1007** [comment]
Updated made. Pushing again

**Review by mdarby — APPROVED** [review]

**Pull request merged**